File: dispatcher/src/main/java/io/guberlo/sapere/dispatcher/Dispatcher.java
package io.guberlo.sapere.dispatcher;

import io.guberlo.sapere.utils.config.Config;
import io.guberlo.sapere.utils.elastic.ElasticUtils;
import io.guberlo.sapere.utils.kafka.KafkaUtils;
import io.guberlo.sapere.utils.spark.SparkUtils;
import org.apache.spark.api.java.function.VoidFunction2;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.functions;
import org.apache.spark.sql.streaming.StreamingQueryException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.concurrent.TimeoutException;


public class Dispatcher {

    private static final Logger LOG = LoggerFactory.getLogger(Dispatcher.class);
    private final KafkaUtils kafkaUtils;
    private final SparkUtils sparkUtils;
    private transient final ElasticUtils elasticUtils;
    private transient final Config config;
    private final String broker;
    private final String topic;
    private final List<String> processTopic;

    public Dispatcher(String lang, String configPath) {
        config = new Config().getYaml(configPath);
        kafkaUtils = new KafkaUtils();
        sparkUtils = new SparkUtils(lang);
        elasticUtils = new ElasticUtils(config.getElasticConfig().getIndex(), config.getElasticConfig().getHost());
        broker = config.getKafkaConfig().getBrokers();
        topic = config.getKafkaConfig().getGeneralTopic();
        processTopic = config.getKafkaConfig().getProcessTopic();
    }

    public Dataset<Row> preProcess(Dataset<Row> dataset) {

        // 1. Convert from Kafka schema to spark and remove NA records
        Dataset<Row> converted = fromKafkaToSpark(dataset);
        if (converted == null) {
            LOG.info("Dataset is empty");
            return null;
        }

        // 2. If specified in the request, tokenize and remove stop words
        // TODO: Define how to establish if the request has those settings
        if (1 == 0)
            return sparkUtils.removeStopWords(sparkUtils.tokenize(converted));

        return converted;
    }

    private void func(Dataset<Row> batch, Long batchId) {
        batch.foreachPartition(elasticUtils::updateOnES);
    }

    /**
     * IMPORTANT: this method is used only when the document is already
     * Indexed on elastic search
     *
     * Parse the received DataFrame into the appropriate schema
     * And apply a function to save data into Elastic Search with
     * The corrisponding predictions.
     *
     * @param output dataset off predictions to send to elastic
     */
    private void updateOnElastic(Dataset<Row> output) throws TimeoutException, StreamingQueryException {
        output.selectExpr("CAST(value AS STRING)")
                .select(functions.from_json(functions.col("value"), sparkUtils.getOutputSchema()).alias("data"))
                .select("data.*")
                .writeStream()
                .foreachBatch((VoidFunction2<Dataset<Row>, Long>) this::func)
                .start()
                .awaitTermination();
    }

    public void start() throws TimeoutException, StreamingQueryException {
        Dataset<Row> stream = kafkaUtils.readKafkaStream(topic, broker);
        stream = preProcess(stream);
        kafkaUtils.writeStreamKafka(stream, processTopic.get(0), broker);

        Dataset<Row> output = kafkaUtils.readKafkaStream(processTopic.get(1), broker);
        updateOnElastic(output);
    }

    private Dataset<Row> fromKafkaToSpark(Dataset<Row> dataset) {
        return dataset.selectExpr("CAST(value AS STRING)")
                .alias("json")
                .select(functions.from_json(functions.col("value"), sparkUtils.getInputSchema()).alias("data"))
                .select("data.*")
                .na().drop();

    }

}

File: PySpark/dispatcher/dispatcher.py
import yaml

from typing import Iterator

from pyspark import SparkContext

from pyspark.sql import SparkSession
from pyspark.sql.dataframe import DataFrame
from pyspark.sql.functions import from_json, udf, to_json, struct
from pyspark.sql import Row
from pyspark.sql.types import *
from pyspark.sql import types as tp

from pyspark.ml.feature import StopWordsRemover, Tokenizer

from KafkaUtils import readStreamKafka, writeStreamKafka # imported using --py-files
from ConfigUtils import getYaml
from SparkUtils import inputSchema, outputSchema, writeToConsole

from elasticsearch import Elasticsearch

cfg = getYaml("/opt/spark/config/config.yaml")

brokers = cfg['kafka']['brokers']
topic = cfg['kafka']['generalTopic']
processTopic = cfg['kafka']['processTopics']

elasticHost = cfg['elastic']['host']
index = cfg['elastic']['index']

context = SparkContext(appName="Requests Dispatcher")
session = SparkSession(context)
lang = "english"

context.setLogLevel("WARN")

tokenizer = Tokenizer(inputCol="text", outputCol="tokens")
countTokens = udf(lambda words: len(words), tp.IntegerType())

stopWords = StopWordsRemover.loadDefaultStopWords(lang)
remover = StopWordsRemover(inputCol="tokens", outputCol="filtered", stopWords=stopWords)

print(f"Spark version = {context.version}")
print(f"Hadoop version = {context._jvm.org.apache.hadoop.util.VersionInfo.getVersion()}")

def preprocessing(stream: DataFrame) -> DataFrame:
    """
        Take JSON data from Kafka and apply some basic NLP,
        if requested.

        Keyword argument:
        stream -- DataFrame containing data from API.
    """
    request = 0

    # Fit DataFrame from kafka to schema specified above and removes NA
    words = stream.selectExpr("CAST(value AS STRING)") \
            .alias("json")
    words = words.select(from_json(words.value, inputSchema).alias("data")) \
            .select("data.*")
    words = words.na.drop()

    if not words:
        print("No Data")
        return 

    # Apply tokenization/stopwords if specified
    if request == 1:
        tokenized = tokenize(words)
        stopRemoved = removeStopWords(tokenized)
        return stopRemoved
    
    return words

def tokenize(words: DataFrame) -> DataFrame:
    """
        Apply tokenization using spark ML feature.
    """
    tokenized = tokenizer.transform(words)
    return tokenized

def removeStopWords(tokens: DataFrame) -> DataFrame:
    """
        Remove stop words using spark ML feature.
    """
    stopRemoved = remover.transform(tokens)
    return stopRemoved

def func(batch_df: DataFrame, batch_id: int) -> None:
    """
        Save each batch into Elastic Search updating the 
        Already existing resource with new predictions.
    """
    rdd = batch_df.foreachPartition(saveToElastic)
    
def saveToElastic(data: Iterator) -> None:
    """
        Update the corresponding Elastic Search resource
        With the corresponding prediction. 
    """
    es = Elasticsearch(hosts=[elasticHost]) # We need to create a new ES because it is not de-serializable and can't be passed to a spark executor
    for doc in data:
        id = doc.id
        es.update(index=index, id=id, body={"doc": {doc.type:doc.prediction}})

def sendToElastic(stream: DataFrame) -> None:
    """
        Parse the received DataFrame into the appropriate schema
        And apply a function to save data into Elastic Search with
        The corrisponding predictions.
    """
    stream.selectExpr("CAST(value AS STRING)") \
        .select(from_json("value", outputSchema).alias("data")) \
        .select("data.*") \
        .writeStream \
        .foreachBatch(func) \
        .start() \
        .awaitTermination()

if __name__ == "__main__":
    # Read stream from kafka to a DataFrame
    stream = readStreamKafka(topic, brokers)
    words = preprocessing(stream)
    words.printSchema()

    writeStreamKafka(words, processTopic[0], brokers)

    out = readStreamKafka(processTopic[1], brokers)
    sendToElastic(out)
